src/services/fetch.rs

```diff
@@ -446,3 +511,197 @@ impl Drop for FetchTask {
+        if let Json(Ok(http_bin)) = resp.body() {
+            assert_eq!(http_bin.url, String::from("https://httpbin.org/get"));
+        } else {
+            assert!(false, "unexpected resp: {:#?}", resp);
+        }
+    }
+
+    #[test]
+    async fn fetch_redirect_follow() {
+        let request = Request::get("https://httpbin.org/relative-redirect/1")
+            .body(Nothing)
+            .unwrap();
+        let options = FetchOptions {
+            redirect: Some(Redirect::Follow),
+            ..FetchOptions::default()
+        };
+        let cb_future = CallbackFuture::<Response<Json<Result<HttpBin, anyhow::Error>>>>::default();
+        let callback: Callback<_> = cb_future.clone().into();
+        let _task = FetchService::new().fetch_with_options(request, options, callback);
+        let resp = cb_future.await;
+        assert_eq!(resp.status(), StatusCode::OK);
+        if let Json(Ok(http_bin)) = resp.body() {
+            assert_eq!(http_bin.url, String::from("https://httpbin.org/get"));
+        } else {
+            assert!(false, "unexpected resp: {:#?}", resp);
+        }
+    }
+
+    #[test]
+    async fn fetch_redirect_error() {
+        let request = Request::get("https://httpbin.org/relative-redirect/1")
+            .body(Nothing)
+            .unwrap();
+        let options = FetchOptions {
+            redirect: Some(Redirect::Error),
+            ..FetchOptions::default()
+        };
+        let cb_future = CallbackFuture::<Response<Result<String, anyhow::Error>>>::default();
+        let callback: Callback<_> = cb_future.clone().into();
+        let _task = FetchService::new().fetch_with_options(request, options, callback);
+        let resp = cb_future.await;
+        assert_eq!(resp.status(), StatusCode::REQUEST_TIMEOUT);
+    }
+
+    #[test]
+    async fn fetch_redirect_manual() {
+        let request = Request::get("https://httpbin.org/relative-redirect/1")
+            .body(Nothing)
+            .unwrap();
+        let options = FetchOptions {
+            redirect: Some(Redirect::Manual),
+            ..FetchOptions::default()
+        };
+        let cb_future = CallbackFuture::<Response<Result<String, anyhow::Error>>>::default();
+        let callback: Callback<_> = cb_future.clone().into();
+        let _task = FetchService::new().fetch_with_options(request, options, callback);
+        let resp = cb_future.await;
+        assert_eq!(resp.status(), StatusCode::OK);
+        // body is empty because the response is opaque for manual redirects
+        assert_eq!(resp.body().as_ref().unwrap(), &String::from(""));
+    }
+
+    #[test]
+    async fn fetch_integrity() {
+        let request = Request::get("https://httpbin.org/base64/WWV3IFNSSSBUZXN0")
+            .body(Nothing)
+            .unwrap();
+        let options = FetchOptions {
+            integrity: Some(
+                "sha384-4xsxy1bN3ru9z4rMCqjTPrPYnN91zxlpAvHO4/fDu9kIxWsxmnbA7F/qFbuTuQDJ"
```

Comment:
Can you please add a test where the integrity doesn't match and so the fetch results in a network error?

Reply:
Added!